keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,815 @@
+- Creation of invalid CRs
+- Reject if a VolumeSnapshot CR does not have a legit VolumeSnapshotSource, i.e., missing both PersistentVolumeName and VolumeSnapshotContentName.
+- Reject if a VolumeSnapshotContent CR does not have a legit VolumeSnapshotContentSource, i.e., both VolumeHandle and SnapshotHandle have been specified
+- Updating immutable fields
+- Reject updates to VolumeSnapshot’s VolumeSnapshotSource
+- Reject updates to VolumeSnapshotContent’s VolumeSnapshotContentSource
+- Reject updates to VolumeSnapshotContent’s volume snapshot ref after binding
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Backwards compatibility
+
+There is a backwards compatibility issue involved when tightening the validation on snapshot objects. Since the feature is already in beta we are committed to more backward compatibility guarantee than alpha.
+
+There are two perspectives.
+- client perspective: API requests that previously succeeded now fail (these are okay)
+  - create: invalid things get rejected rather than persisted
+  - update: invalid transitions now get rejected
+- API server perspective for existing persisted data
+  - do not tighten validation (via schema or webhook) in a way that would prevent a no-op update (this must not be violated)
+  - reason: removing a finalizer is done via update, causes problems with delete
+  - making a previously optional field required in the schema blocks update of previously persisted data that omitted the field (unless the update populates the newly required field, or you specify a schema default)
+
+To tackle the backwards compatibility problem, this KEP proposes the following release process.
+
+Begin with validating webhook only enforcement. The webhook will perform the following validation
+- one release with ratcheting validation using the webhook server
+   - webhook is strict on create
+   - webhook is strict on updates where the existing object passes strict validation
+   - webhook is relaxed on updates where the existing object fails strict validation (allows finalizer removal, status update, deletion, etc)
+ - some reconciliation process in volume snapshot controller that ensures invalid data is fixed or removed. This process will be entirely user managed. The webhook and controllers will not take any automatic action to reconcile invalid objects.
```

Comment:
Can you clarify this? How does the snapshot controller ensure that invalid objects are fixed/removed without automatically reconciling anything and letting user control it?

Reply:
The snapshot controller will not do anything. Added a note to clarify.